Build failures with Emscripten image > 3.1.12 (when building from source)

@kelson42 @mossroy I've tried to reproduce a build based on the instructions in the README.

I encountered two errors. The first was easy to fix: zlib v1.2.12 no longer exists, so I had to update the reference to it in Makefile to v1.2.13.

The second is much harder (for me) to fix. I consistently get the following error(s) -- I tried three times -- when running the docker / make command, which occurs towards what I believe is the end of the build (it takes a very long time to get to that point before it bombs out):
```
/emsdk/upstream/emscripten/cache/sysroot/include/c++/v1/__memory/uninitialized_algorithms.h:628:21: 
error: use of overloaded operator '!=' is ambiguous (with operand types 'std::reverse_iterator<zim::offset_t *>' and 'std::reverse_iterator<zim::offset_t *>')
    while (__first1 != __last1) {
```
Easier to see in screenshot:

<img width="856" alt="image" src="https://user-images.githubusercontent.com/4304337/202652902-2593837b-dcb8-4529-a243-27e458b2232c.png">

This is still an issue when building libzim from source rather than from pre-compiled binaries. We can build using Enscripten emsdk 3.1.41 (current latest) if building using a pre-compiled binary, but building from source still requires 3.1.12, due to the error:

```
/emsdk/upstream/emscripten/cache/sysroot/include/c++/v1/__memory/uninitialized_algorithms.h:632:21: error: use of
overloaded operator '!=' is ambiguous (with operand types 'std::reverse_iterator<zim::offset_t *>' and
'std::reverse_iterator<zim::offset_t *>')
  632 |     while (__first1 != __last1) {
      |            ~~~~~~~~ ^  ~~~~~~~
```

For the relevant workflow run, see https://github.com/openzim/javascript-libzim/actions/runs/5351035516/jobs/9704382137.

--- a/.github/workflows/build_libzim_wasm.yml
+++ b/.github/workflows/build_libzim_wasm.yml
@@ -53,7 +53,7 @@ jobs:
     # Customizes the Emscripten docker container via the Dockerfile in this repo
     - name: Build the Docker image
       run: |
-        if [[ $BUILD_TYPE -eq 'source' ]]; then
+        if [[ $BUILD_TYPE =~ 'source' ]]; then
           BUILD_VERSION='3.1.12'
         else
           BUILD_VERSION='3.1.41'

--- a/.github/workflows/build_libzim_wasm.yml
+++ b/.github/workflows/build_libzim_wasm.yml
@@ -42,6 +42,7 @@ env:
   DISPATCH_TYPE: ${{ github.event.inputs.buildtype }}
   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
   SSH_KEY: ${{ secrets.SSH_KEY }}
+  BUILD_TYPE: ${{ github.event.inputs.buildtype }}
 
 jobs:
   build:
@@ -51,7 +52,13 @@ jobs:
     - uses: actions/checkout@v3
     # Customizes the Emscripten docker container via the Dockerfile in this repo
     - name: Build the Docker image
-      run: docker build -t "docker-emscripten-libzim:v3" ./docker
+      run: |
+        if [[ $BUILD_TYPE -eq 'source' ]]; then
+          BUILD_VERSION='3.1.12'
+        else
+          BUILD_VERSION='3.1.41'
+        fi
+        docker build -t "docker-emscripten-libzim:v3" ./docker --build-arg VERSION=$BUILD_VERSION
     # If we're building release version
     - name: Build release from libzim binaries
       if: github.event_name == 'pull_request' || github.event_name == 'push' || github.event.inputs.buildtype == 'release'

--- a/docker/Dockerfile
+++ b/docker/Dockerfile
@@ -1,5 +1,6 @@
 # Docker image suitable to build libzim & kiwixlib with emscripten
-FROM emscripten/emsdk:3.1.41
+ARG VERSION
+FROM emscripten/emsdk:${VERSION}
 
 # Install build tools
 RUN apt-get update && apt-get install -y --no-install-recommends wget ninja-build pkg-config python3 autopoint libtool autoconf patch g++ locales curl git \

